A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,529 @@
+RPCs; instead, they should be careful to set status codes carefully on
+data plane RPCs on both the client and server side.  In most cases, it
+is expected that UNAVAILABLE is an appropriate code for data plane RPCs.
+
+### Temporary environment variable protection
+
+The new functionality will be guarded by the
+`GRPC_EXPERIMENTAL_XDS_DATA_ERROR_HANDLING` env var until it passes
+interop testing.  This env var will guard using the new response fields
+from [xRFC TP3] and the new server features "fail_on_data_errors" and
+"resource_timer_is_transient_error".  It will also guard whether we
+default to ignoring resource deletion.
+
+## Rationale
+
+We considered an approach whereby we'd notify the watchers about whether
+the error is a data error or a transient error and let the watcher make
+the decision about how to respond.  However, that would make the watcher
+API much harder to understand.  It would also break CSDS and XdsClient
+metrics in that it would divorce the cache status inside the XdsClient
+from the watcher behavior outside of the XdsClient.
+
+## Implementation
+
+C-core implementation:
+- XdsClient watcher API change: https://github.com/grpc/grpc/pull/38269
+- resolution_note handling: https://github.com/grpc/grpc/pull/38341 and
+  https://github.com/grpc/grpc/pull/38411
+- Support for "fail_on_data_errors": https://github.com/grpc/grpc/pull/38278
+- Support for "resource_timer_is_transient_failure": https://github.com/grpc/grpc/pull/38377
+- xRFC TP3 support: https://github.com/grpc/grpc/pull/38381
+
+Will also be implemented in Java, Go, and Node.
+
+## Open issues
+
+To avoid problems where newly started clients fail when the control plane
+is down, we have designed xDS fallback functionality, as described in
+[gRFC A71](A71-xds-fallback.md).  However, because we do not want to
+trigger fallback due to an individual resource not existing if the control
+plane is actually responding, we do not currently trigger fallback when
+the resource timer fires if we are still connected to the control plane.
+This has led to concerns about failure modes where control plane slowness
+may cause clients to act as if resources do not exist rather than using a
+fallback server.  Ideally, it seems like that we would like to use a timer
+to consider the xDS server connection to have failed if we haven't gotten
+back a response in a certain amount of time, but we cannot do that while
+we are depending on the timer to detect individual resources not existing.
+
+As future work, we should consider whether there are ways to safely
+trigger xDS fallback if the primary server is responding too slowly.
+There are a number of challenges here:
+
+- Given the number of issues we've seen where the control plane has been
+  slow to respond, it's not clear to me that we would want all of those
+  cases to trigger fallback, since we'd likely hit fallback way too often.
+  I suspect that we would first need the control plane to monitor and provide
```

Comment:
s/I/We/?   And one instance at the bottom, too.

Reply:
Done.